python/dgl/distributed/kvstore.py

```diff
@@ -0,0 +1,1008 @@
+    """
+    def __init__(self, msg):
+        self.msg = msg
+
+    def __getstate__(self):
+        return self.msg
+
+    def __setstate__(self, state):
+        self.msg = state
+
+class InitDataRequest(rpc.Request):
+    """Send meta data to server and init data tensor
+    on server using UDF init function.
+
+    Parameters
+    ----------
+    name : str
+        data name
+    shape : tuple
+        data shape
+    dtype : str
+        data type string, e.g., 'int64', 'float32', etc.
+    policy_str : str
+        partition-policy string, e.g., 'edge' or 'node'.
+    init_func : function
+        UDF init function.
+    """
+    def __init__(self, name, shape, dtype, policy_str, init_func):
+        self.name = name
+        self.shape = shape
+        self.dtype = dtype
+        self.policy_str = policy_str
+        self.init_func = init_func
+
+    def __getstate__(self):
+        return self.name, self.shape, self.dtype, self.policy_str, self.init_func
+
+    def __setstate__(self, state):
+        self.name, self.shape, self.dtype, self.policy_str, self.init_func = state
+
+    def process_request(self, server_state):
+        kv = server_state.kv_store
+        dtype = F.data_type_dict[self.dtype]
+        if kv.is_backup_server() == False:
+            data_tensor = self.init_func(self.shape, dtype)
+            kv.init_data(name=self.name, data_tensor=data_tensor)
+        else: # backup server will read data from shared-memory
+            kv.init_data(name=self.name)
+        # Find the same partition policy from exsiting plolicy list.
+        for _, policy in kv.part_policy.items():
+            if policy.policy_str == self.policy_str:
+                kv.part_policy[self.name] = policy
+                res = InitDataResponse(INIT_MSG)
+                return res
+        raise RuntimeError("Cannot find any partition policy match \
+            the requested policy : %s" % self.policy_str)
+
+BARRIER = 901234
+BARRIER_MSG = 'Barrier'
+
+class BarrierResponse(rpc.Response):
+    """Send an confimation signal (just a short string message) of
+    BarrierRequest to client.
+
+    Parameters
+    ----------
+    msg : string
+        string msg
+    """
+    def __init__(self, msg):
+        self.msg = msg
+
+    def __getstate__(self):
+        return self.msg
+
+    def __setstate__(self, state):
+        self.msg = state
+
+class BarrierRequest(rpc.Request):
+    """Send a barrier signal (just a short string message) to server.
+
+    Parameters
+    ----------
+    msg : string
+        string msg
+    """
+    def __init__(self, msg):
+        self.msg = msg
+
+    def __getstate__(self):
+        return self.msg
+
+    def __setstate__(self, state):
+        self.msg = state
+
+    def process_request(self, server_state):
+        assert self.msg == BARRIER_MSG
+        kv = server_state.kv_store
+        kv.incr_barrier_count()
+        if kv.barrier_count == kv.num_clients:
+            kv.barrier_zero()
```

Comment:
maybe you can move this logic to kvstore so you don't need to create this kind of functions.